.github/workflows/manual-upgrade-testnet-l2.yml

```
@@ -135,11 +130,25 @@ jobs:
         with:
           creds: ${{ secrets.AZURE_CREDENTIALS }}
 
+      - name: 'Fetch latest VM hostnames by env tag and extract build number'
+        id: fetch_hostnames
+        run: |
+          TAG_NAME="${{needs.build.outputs.RESOURCE_TAG_NAME}}"
+          VM_HOSTNAME=$(az vm list --query "[?tags.$TAG_NAME=='true'].{Name:name}[0]" -o tsv)
+          VM_BUILD_NUMBER=$(echo $VM_HOSTNAME | perl -ne 'if (/(-[0-9]{1}-)(\d+)/) { print $2 }') # Extract build number from VM hostname, e.g. D-0-321 -> 321
+          echo "VM_BUILD_NUMBER=VM_BUILD_NUMBER" >> $GITHUB_ENV
+          if ! [[ $BUILD_NUMBER =~ ^[0-9]+$ ]]; then # fail if build number is not a number
+            echo "Error: Hostname lookup or regex extraction of build number failed."
+            echo "VM_HOSTNAME: ${VM_HOSTNAME}"
+            echo "VM_BUILD_NUMBER: ${VM_BUILD_NUMBER}"
+            exit 1
+          fi
```

Would print the values even if not error.

```suggestion
          echo "VM_BUILD_NUMBER=VM_BUILD_NUMBER" >> $GITHUB_ENV
          echo "VM_HOSTNAME: ${VM_HOSTNAME}"
          echo "VM_BUILD_NUMBER: ${VM_BUILD_NUMBER}"
          if ! [[ $BUILD_NUMBER =~ ^[0-9]+$ ]]; then # fail if build number is not a number
            echo "Error: Hostname lookup or regex extraction of build number failed."
            exit 1
          fi
```